### PR TITLE
Improve dependencies completion in Cargo.toml

### LINKED ISSUE #2856: Cargo.toml crate completion doesn't work for target-specific dependencies
## Environment

* **Intellij-Rust plugin version:** 0.2.0.2104-181
* **Rust toolchain version:** rustc 1.28.0 (9634041f0 2018-07-30)
* **IDE name and version:** IntelliJ IDEA 2018.1.5 (Community Edition)
* **Operating system:** Arch Linux

## Problem description

Completion of crates.io crates doesn't seem to work inside target-specific dependency keys like `[target.'cfg(windows)'.dependencies]`

## Steps to reproduce

Try typing crate names in this Cargo.toml:

```toml
[package]
name = "anything"
version = "0.1.0"
authors = ["Jonas Schievink <jonasschievink@gmail.com>"]

[dependencies]

[dev-dependencies]

[target.'cfg(windows)'.dependencies]

[target.'cfg(unix)'.dependencies]
```

The first 2 dependency keys (`dependencies` and `dev-dependencies`) work fine, the last 2 don't.

Presumably, this is caused by this line, which only looks at the first key (here, `target`) to determine whether completion should be done:

https://github.com/intellij-rust/intellij-rust/blob/6a24958d7dcd86b28edbfd0ad29da2f169fe564b/src/main/kotlin/org/rust/toml/RsTomlCompletionProvider.kt#L56

Instead of looking at the first element, it should probably check the *last* one.

### PR DESCRIPTION
**Pull request by vlad20012**
With this PR we should be able to complete versions of dependencies in cargo toml when caret is placed to toml value position, i.e.
```
[dependencies]
rand <caret>
rand = <caret>
rand = "<caret>"

[dependencies.<caret>]
<caret>
```

Fixes #2856

### CODE REVIEW

**Comment by vlad20012** [comment]
Ready for review

File: src/main/kotlin/org/rust/toml/CratesIoApi.kt [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    val name: String,
+    @SerializedName("max_version")
+    val maxVersion: String
+)
+
+val CrateDescription.dependencyLine: String
+    get() = "$name = \"$maxVersion\""
+
+fun searchCrate(key: TomlKey): Collection<CrateDescription> {
+    MOCK?.let { return it }
```
Comment (Undin) [inline src/main/kotlin/org/rust/toml/CratesIoApi.kt]: I like the way how we implement other mocks, for example, for extract function refactoring
More precisely, I like the error if `MOCK` variable in tests is null and error message.
Such error prevents you from accidental errors in tests and message says exactly what you should do
It would be great use such pattern here as well

File: src/main/kotlin/org/rust/toml/CratesIoApi.kt [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+private inline fun <reified T> requestCratesIo(context: PsiElement, path: String): T? {
+    return requestCratesIo(context, path, T::class.java)
+}
+
+private fun <T> requestCratesIo(context: PsiElement, path: String, cls: Class<T>): T? {
+    return try {
+        runWithCheckCanceled {
+            val response = HttpRequests.request("https://crates.io/api/v1/$path")
+                .readString(ProgressManager.getInstance().progressIndicator)
```
[the remark this answers is not in the document]
Comment (Undin) [reply src/main/kotlin/org/rust/toml/CratesIoApi.kt]: Okay, `HttpURLConnection` can reuse connections, but I still don't like it because of other comments))

File: src/main/kotlin/org/rust/toml/CargoTomlDependenciesCompletionProvider.kt [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        }
+    }
+
+    protected abstract fun completeDependencyKey(key: TomlKey, result: CompletionResultSet)
+
+    protected abstract fun completeDependencyValue(keyValue: TomlKeyValue, result: CompletionResultSet)
+}
+
+/** @see CargoTomlPsiPattern.inDependencyKeyValue */
+class CargoTomlDependenciesCompletionProvider : AbstractTomlKeyValueCompletionProvider() {
```
Comment (Undin) [inline src/main/kotlin/org/rust/toml/CargoTomlDependenciesCompletionProvider.kt]: Let's use the same naming style for these providers and replace `Dependencies` to `Dependency`

File: src/main/kotlin/org/rust/toml/CargoTomlDependenciesCompletionProvider.kt [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+        }
+
+        if (!hasQuotes) {
+            context.document.insertString(startOffset, "\"")
+            context.document.insertString(context.selectionEndOffset, "\"")
+        }
+    }
+}
+
+private inline fun <reified T : PsiElement> InsertionContext.getItemOfType(strict: Boolean = false): T? =
```
Comment (Undin) [inline src/main/kotlin/org/rust/toml/CargoTomlDependenciesCompletionProvider.kt]: Let's drop this function and use `org.rust.lang.core.completion.LookupElementsKt#getElementOfType` instead

File: src/main/kotlin/org/rust/toml/CargoTomlDependenciesCompletionProvider.kt [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+            ?: error("PsiElementPattern must not allow KeyValues in tables without header")
+    }
+}
+
+/** Inserts `=` between key and value if missed and wraps inserted string with quotes if needed */
+private class StringValueInsertionHandler(val keyValue: TomlKeyValue) : InsertHandler<LookupElement> {
+    override fun handleInsert(context: InsertionContext, item: LookupElement) {
+        var startOffset = context.startOffset
+        val value = context.getItemOfType<TomlValue>()
+        val hasEq = keyValue.children.find { it.elementType == TomlElementTypes.EQ } != null
```
Comment (Undin) [inline src/main/kotlin/org/rust/toml/CargoTomlDependenciesCompletionProvider.kt]: `find { someCondition } != null` -> `any { someCondition }`?

File: src/main/kotlin/org/rust/toml/CargoTomlPsiPattern.kt [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+                        val key = context?.get(TOML_KEY_CONTEXT_NAME) ?: return@with false
+                        val names = header.names
+                        names.getOrNull(names.size - 2)?.isDependencyKey == true && names.lastOrNull() == key
+                    }
+            )
+
+    /** Any element inside [onSpecificDependencyHeaderKey] */
+    val inSpecificDependencyHeaderKey: PsiElementPattern.Capture<PsiElement> =
+        cargoTomlPsiElement<PsiElement>()
+            .withSuperParent(1, onSpecificDependencyHeaderKey)
```
Comment (Undin) [inline src/main/kotlin/org/rust/toml/CargoTomlPsiPattern.kt]: `withSuperParent(1, onSpecificDependencyHeaderKey)` -> `withParent(onSpecificDependencyHeaderKey)`

File: src/main/kotlin/org/rust/toml/CargoTomlPsiPattern.kt [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+     * regex = "1"
+     *       ^
+     * ```
+     */
+    private val onDependencyKeyValue: PsiElementPattern.Capture<TomlKeyValue> =
+        cargoTomlPsiElement<TomlKeyValue>()
+            .withSuperParent(
+                1,
+                onDependencyTable
+            )
```
Comment (Undin) [inline src/main/kotlin/org/rust/toml/CargoTomlPsiPattern.kt]: `withParent(onDependencyTable)`

File: src/main/kotlin/org/rust/toml/CargoTomlPsiPattern.kt [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+     * version = "1"
+     *         ^
+     * ```
+     */
+    private val onSpecificDependencyKeyValue: PsiElementPattern.Capture<TomlKeyValue> =
+        cargoTomlPsiElement<TomlKeyValue>()
+            .withSuperParent(
+                1,
+                onSpecificDependencyTable
+            )
```
Comment (Undin) [inline src/main/kotlin/org/rust/toml/CargoTomlPsiPattern.kt]: `withParent(onSpecificDependencyTable)`

File: src/test/kotlin/org/rust/toml/CargoTomlPsiPatternTest.kt [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+    fun `test inSpecificDependencyKeyValue 3`() = testPatternNegative(inSpecificDependencyKeyValue, """
+        [dependencies]
+        version = ""
+        #^
+    """)
+
+    private inline fun <reified T : PsiElement> testPattern(pattern: ElementPattern<T>, @Language("Toml") code: String) {
+        InlineFile(code, "Cargo.toml")
+        val element = findElementInEditor<T>()
+        LightPlatformCodeInsightFixtureTestCase.assertTrue(pattern.accepts(element))
```
Comment (Undin) [inline src/test/kotlin/org/rust/toml/CargoTomlPsiPatternTest.kt]: Looks like you can omit `LightPlatformCodeInsightFixtureTestCase` here

File: src/test/kotlin/org/rust/toml/CargoTomlPsiPatternTest.kt [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+    }
+
+    private fun <T> testPatternNegative(pattern: ElementPattern<T>, @Language("Toml") code: String) {
+        InlineFile(code, "Cargo.toml")
+        val element = findElementInEditor<PsiElement>()
+        LightPlatformCodeInsightFixtureTestCase.assertFalse(pattern.accepts(element))
```
Comment (Undin) [inline src/test/kotlin/org/rust/toml/CargoTomlPsiPatternTest.kt]: The same here

File: src/test/kotlin/org/rust/toml/CargoTomlDependenciesCompletionTest.kt [hunk 11]
```diff
@@ -0,0 +1,164 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.toml
+
+import org.intellij.lang.annotations.Language
+
+class CargoTomlDependenciesCompletionTest : CargoTomlCompletionTestBase() {
```
Comment (Undin) [inline src/test/kotlin/org/rust/toml/CargoTomlDependenciesCompletionTest.kt]: Now dependency completion works with complex keys like `target.'cfg(windows)'.dependencies`  and `target.'cfg(windows)'.dependencies.dep`.
But we still don't have the corresponding tests. Let's add them.

File: src/main/kotlin/org/rust/toml/CargoTomlDependenciesCompletionProvider.kt [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+import com.intellij.psi.tree.IElementType
+import com.intellij.psi.util.PsiTreeUtil
+import com.intellij.util.ProcessingContext
+import org.rust.lang.core.psi.ext.ancestorOrSelf
+import org.rust.lang.core.psi.ext.ancestorStrict
+import org.rust.lang.core.psi.ext.elementType
+import org.rust.lang.core.psi.ext.isAncestorOf
+import org.toml.lang.psi.*
+
+abstract class AbstractTomlKeyValueCompletionProvider : CompletionProvider<CompletionParameters>() {
```
[the remark this answers is not in the document]
Comment (vlad20012) [reply src/main/kotlin/org/rust/toml/CargoTomlDependenciesCompletionProvider.kt]: Partially resolved: I don't think that the name should contains "Dependency" because this class can be used for any TomlKeyValue completion. I removed any mentions of "Dependency" word from this class

File: src/test/kotlin/org/rust/toml/CargoTomlDependenciesCompletionTest.kt (hunk 11 above)
Comment (vlad20012) [reply src/test/kotlin/org/rust/toml/CargoTomlDependenciesCompletionTest.kt]: These cases are covered by `CargoTomlPsiPatternTest`. Should be double-check it?

Comment (Undin) [reply src/test/kotlin/org/rust/toml/CargoTomlDependenciesCompletionTest.kt]: But it's not really true.
Yeah, we indeed have tests for patterns. But are we sure that we correctly use these patterns in other code to make completion works in a necessary way?
In my opinion, pattern tests are convenient way for developers to find out a problem when tests fail. User know nothing about patterns but he wants that completion works. So, first of all, we should test the user use case and other tests are optional/auxiliary

Comment (vlad20012) [reply src/test/kotlin/org/rust/toml/CargoTomlDependenciesCompletionTest.kt]: Ok, I've added this test:
```kotlin
fun `test empty key with complex dependency head`() = doTest("""
        [target.'cfg(windows)'.dev-dependencies]
        <caret>
    """, """
        [target.'cfg(windows)'.dev-dependencies]
        dep = "1.0"
    """, "dep" to "1.0")
```

Comment (Undin) [reply src/test/kotlin/org/rust/toml/CargoTomlDependenciesCompletionTest.kt]: 👍
